dali/kernels/slice/slice_gpu.cuh

```diff
@@ -36,78 +36,122 @@ template <int Dims>
     return;
   }
 
   for (; offset < block_end; offset += blockDim.x) {
-    size_t idx = offset;
-    size_t out_idx = idx;
-    size_t in_idx = 0;
+    int64_t idx = offset;
+    int64_t out_idx = idx;
+    int64_t in_idx = 0;
+    int i_c = 0;
+    bool out_of_range = false;
     for (int d = 0; d < Dims; d++) {
       int i_d = idx / out_strides[d];
-      idx %= static_cast<size_t>(out_strides[d]);
-      in_idx += i_d * static_cast<size_t>(in_strides[d]);
+      idx %= out_strides[d];
+      if (d == channel_dim)
+        i_c = i_d;
+
+      int in_i_d = anchor[d] + i_d;
+      out_of_range |= in_i_d < 0 || in_i_d >= in_shape[d];
+      if (!out_of_range)
+        in_idx += in_i_d * in_strides[d];
     }
     in_idx += idx;  // remaining dims have equal strides
-    out[out_idx] = clamp<OutputType>(in[in_idx]);
+    out[out_idx] = out_of_range ? fill_values[i_c] : clamp<OutputType>(in[in_idx]);
   }
 }
 
 template <typename OutputType, typename InputType, int Dims>
 __global__ void SliceKernel(const SliceSampleDesc<Dims> *samples, const BlockDesc *blocks) {
   int sampleIdx = blocks[blockIdx.x].sampleIdx;
-  size_t offset = blocks[blockIdx.x].offset + threadIdx.x;
-  size_t block_end = blocks[blockIdx.x].offset + blocks[blockIdx.x].size;
+  int64_t offset = blocks[blockIdx.x].offset + threadIdx.x;
+  int64_t block_end = blocks[blockIdx.x].offset + blocks[blockIdx.x].size;
   auto sample = samples[sampleIdx];
   auto *out = static_cast<OutputType*>(sample.out);
   auto *in = static_cast<const InputType*>(sample.in);
-  SliceFunc<Dims>(out, in, sample.out_strides.data(), sample.in_strides.data(), offset, block_end);
+  auto *fill_values = static_cast<const OutputType*>(sample.fill_values);
+  SliceFunc<Dims>(out, in, sample.out_strides.data(), sample.in_strides.data(),
+                  sample.anchor.data(), sample.in_shape.data(), fill_values, sample.channel_dim,
+                  offset, block_end);
 }
 
 }  // namespace detail
 
 template <typename OutputType, typename InputType, int Dims>
 class SliceGPU {
  private:
-  static constexpr size_t kBlockDim = 256;
-  static constexpr size_t kBlockSize = 64 * kBlockDim;
-  size_t block_count_ = 0;
+  static constexpr int64_t kBlockDim = 256;
+  static constexpr int64_t kBlockSize = 64 * kBlockDim;
+  int64_t block_count_ = 0;
 
  public:
   KernelRequirements Setup(KernelContext &context,
                            const InListGPU<InputType, Dims> &in,
-                           const std::vector<SliceArgs<Dims>> &slice_args) {
+                           const std::vector<SliceArgs<OutputType, Dims>> &slice_args) {
     KernelRequirements req;
     ScratchpadEstimator se;
-    const size_t num_samples = in.size();
+    auto num_samples = in.size();
     se.add<detail::SliceSampleDesc<Dims>>(AllocType::Host, num_samples);
     se.add<detail::SliceSampleDesc<Dims>>(AllocType::GPU, num_samples);
 
-    std::vector<size_t> sample_sizes;
+    nfill_values_ = 0;
```

Comment:
You should check https://github.com/NVIDIA/DALI/blob/master/dali/operators/image/paste/paste.cu#L45. There is a nice trick used for filling.

Reply:
If I understand it correctly, the trick is there to avoid `%` operator. In our case the channel index is calculated anyway (because we need to transform the index from the output coordinates to the input ones), so I don't think we could benefit from something like this